Added additional toggles

petervannes opened the pull request:
Added options to toggle the default fields “class” and “logger_name” on
on/off

petervannes commented:
Hi Philip,


I do not you response fully, in my pull request i added the option to disable logging the class and thread properties from the json logmessage instead of enabling it.
I added these options because i used the slf4j-json-logger for logging specific states of SOA composites and Fault states from a Fault Policy. The class and thread properties are irrelevant when logging these faults.
I use the slf4j-json-logger in my java class because the default loggers for Oracle SOA products are only capable of logging in a proprietary format which is hard to parse by Splunk where JSON format works very well. Although it does not hurt if the class and thread are logged, less is more when logging in high volumes.



> On 2 Oct 2016, at 17:46, Philip A. Chapman <notifications@github.com> wrote:
> 
> @pachapman approved this pull request.
> 
> Changes look good, though I'm curious as to why you need to show class and thread for what is essentially a utility for a testing framework.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/savoirtech/slf4j-json-logger/pull/2#pullrequestreview-2450814>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHJvnCQXiFdRfVRkUxedYjXPxxWB_9x7ks5qv9HkgaJpZM4KMANK>.
> 



Met vriendelijke groet/With kind regards

Peter van Nes
Mobile +31 652 694 792 | Fax +31 84 747 6999
 <http://twitter.com/petervannes> <http://nl.linkedin.com/in/petervannes> <http://www.reddipped.com/> <skype:peter_van_nes?call> <xmpp:?jid=pcvnes@gmail.com> <mailto:mailbox@petervannes.nl>

The Future Group <http://thefuturegroup.com/>, Dé Maat in ondernemen.

artnaseef commented:
This change looks good.  Thank you for the contribution.

One thing to address with this update - since this changes some of the public API in a non-backward compatible way, we need to bump the major version.  I'm referring specifically to the constructor argument change.

It would be great if we could make this interface less brittle to future changes.  Now would be a good time to do so since we're going to need a new major version here anyway.

Perhaps using setters instead of constructor arguments?  Or, is there some way we could tie into the SLF4J configuration system to get custom configuration here?

pachapman commented:
Why  not just add constructors with default values that mimic the original functionality?  That way we do not lose backward compatibility.

jwcarman commented:
Yeah, just default it to showing the information and let you use the other constructor to turn it off.  +1

artnaseef commented:
I used to be more in-line with adding more code for backward compatibility.

These days, though, I tend to look more heavily at the cost of that tech debt, maintaining that added complexity with no real gain in value, and which can make the code harder-to-use due to the non-obvious method variations.

In this case, I have little concern with bumping the major version.  But, I would like to apply the "Use Before Reuse" concept here - this is the second time we're doing the same (changing the constructor arguments and introducing a breaking change), so it's a good time to solve the reusability in a way that reduces the need for breaking, major version changes.

Another consideration - a setter-based approach is more "Real Obvious" - making the code easier to use, and easier to maintain since we avoid constructor-variation-explosion as subclasses as created over time with their own additional constructor arguments.  It's also more obvious since the setter names are clear, as opposed to matching constructor arguments by-position.

Finally, when I think of Logging and configuration of what's included in the log output, the good-'ol log4j configuration files come to mind.  Having the application hard-code these settings in the long run may not be ideal - especially for anyone using this code in common (i.e. heavily-reused) libraries.  Hence the desire to latch onto an existing Logging platform means of configuration.

petervannes commented:
It was not my intention to make a breaking change, so reverted the constructor and added new attributes with getters/setters. 

Maybe it is an good idea to remove the includeLoggerName from the constructor in the next major release and replace it with a getter/setter. 

Additionally i changed the class name of the JUnit test BasicLoggingTests.java to BasicLoggingTest.java. In Netbeans only JUnit test classes matching the mask
*Test.java are fired.

jwcarman commented:
If you're going to evolve the API and you want to protect yourself from changes to things like constructors, consider using the "builder" pattern instead.

artnaseef commented:
I'm going to merge this PR and then introduce the builder pattern, as suggested by James, and finally release a new major version.

artnaseef commented:
One thing I noticed while working on the builder pattern that I hadn't noticed before - the addition of getters and setters to the JsonLogger interface.

@petervannes - is there a real need to have these methods in the interface, outside of the library itself?

Please let me know as I am in the process of removing them and don't want to miss a need.

artnaseef commented:
Note that I put up a new Pull Request with the builder pattern.  Feedback would be appreciated.
